### ISSUE TITLE
Execute last recorded macro (and map it to Q)

### ISSUE THREAD

**Issue by clason** [neovim/neovim#15404]
Vim has `@@` to repeat the last **executed** macro. This means you currently have to 
1. `qa`, do something, `q`
2. `@a`
3. `@@`
4. repeat
to apply the mapping multiple times. This is inconvenient; there should be a function that executes the last **recorded** macro, which would allow skipping step 2.

This new function could be mapped to `Q`, which right now switches to Ex mode, whose only purpose is to provide new users who have mastered the art of exiting vim with a new challenge. (Ex mode would remain accessible through the alternate mapping `gQ`.)

**Comment by vigoux** [comment]
How would you enter `Ex` mode after that ? I am not a big fan of this mapping myself, but I fear that people using Ex mode would lose it.

**Comment by clason** [comment]
`gQ` still works

**Comment by vigoux** [comment]
Indeed, never knew this existed, thus I am strongly in favor, I don't see the point of having two mapping to do one thing anyway.

**Comment by matveyt** [comment]
It's true that one of `Q` and `gQ` could do something more useful. But I don't see how mapping two (rather rarely used) letters to one is so important. Also, that will probably lead to `qqxxxq` being treated differently from `let @q="xxx"`. Is it really okay?

>One could simply map Q to @@, but this only works if the macro has been executed once; it'd be nicer if it worked even without manual execution first.

Well, why not change the definition of `@@` then?

**Comment by clason** [comment]
The main issue here is that in 99% of cases `Q` will do something _very_ unexpected and annoying, and it'd be much preferable to do something useful instead.

> Also, that will probably lead to qqxxxq being treated differently from let @q="xxx". Is it really okay?

No, why? `Q` is already mapped, so why should there be any difference in your example?

> Well, why not change the definition of @@ then?

Because that (needlessly) breaks previous behavior, and because the definition of `@@` is not the point here.

**Comment by matveyt** [comment]
>it'd be much preferable to do something useful (and easily undoable) instead.

Sure, but why (improved) `@@`? Isn't there something even more useful? Personally, I rarely use `:h complex-repeat` at all. This is why I'm so sceptical.

**Comment by clason** [comment]
Why not? If you have a better idea, you're welcome to propose it -- that's the point of this discussion.

**Comment by matveyt** [comment]
At the very least Vim has `:h gq` (set in defaults.vim). But quite probably every one out there already has its own binding for Q.

**Comment by clason** [comment]
I still don't get your point. Are you arguing that we should leave `Q` and instead map `gQ` to something else?

**Comment by matveyt** [comment]
I say that Vim maps `Q` to `gq` (text format) which is still more useful, IMO, than `@@`.

**Comment by justinmk** [comment]
> A more broadly useful function could be to execute the last macro

Seems reasonable, would this go in a new default register? Or maybe it should always execute `@q`, then `Q` is not a "moving target".

In any case, `reg_recorded()` is missing (counterpart to `reg_executing()`, `reg_recording()`) and would be useful.

> I say that Vim maps `Q` to `gq` (text format) which is still more useful, IMO, than `@@`.

`gq` is already easy to type. A minor alias is not a great use of `Q`.

**Comment by clason** [comment]
> Seems reasonable, would this go in a new default register? Or maybe it should always execute `@q`, then `Q` is not a "moving target".

I think "execute last recorded macro" like `@@` means "repeat last executed macro" would be the most convenient option, but if that is not feasible, `Q` as `@q` is a nice idea since `q` is the "natural throw-away" macro.

(In fact, *yoink*)

**Comment by vigoux** [comment]
I'd rather go with the equivalent of `nnoremap Q @@` here, as tying `Q` to `@q` seems not really convenient.

If I understand correctly, the alternative would be that `Q` would replay the _last recorded macro_ ?

**Comment by justinmk** [comment]
> I think "execute last recorded macro" like `@@` means "repeat last executed macro" would be the most convenient option, but if that is not feasible, `Q` as `@q` is a nice idea since `q` is the "natural throw-away" macro.

Not sure if I was clear. Currently there is unnecessary friction between recording a macro and executing it: having to execute it once is an extra step. The idea is to record a macro and then be able to repeat it, without the extra step.

```
qqfoo<esc>q
Q
Q
Q
```

instead of:

```
qqfoo<esc>q
@q
Q
Q
```

**Comment by vigoux** [comment]
Then I agree that having `Q` replay the last recorded macro is great.

And, to be clear, it is better for it to be _the last recorded macro_ instead of just `@q`.

**Comment by justinmk** [comment]
> OTOH, this could be achievable with scripting, via `reg_recording()`, if there was a `Recording` event...

IMO this is higher priority than deciding what to do with Q.

**Comment by vigoux** [comment]
What would be the best solutions then:
- Add a new read-only register that is the last recorded macro ?
- Create a `Recording` event ?

**Comment by clason** [comment]
Option 2 has probably additional uses (even if I can't think of any off the top of my head), so I'd vote for that.

(Teach a man to record fish, and all that...)

**Comment by vigoux** [comment]
Maybe we could go with both ? I feel like having the last recorded macro is great too.

**Comment by clason** [comment]
(Should this be a new issue, or should I just edit this issue to put the new functionality front an center?)

**Comment by vigoux** [comment]
Keep the discussion, and just change the title

**Comment by justinmk** [comment]
> * Add a new read-only register that is the last recorded macro ?

Maybe, but it's not technically necessary if there is a `Recording` event. 

Could also avoid the register name discussion by adding `reg_recorded()`.

**Comment by vigoux** [comment]
Working on it btw, will try to open a PR asap with both `reg_recorded()` and two new autocmds: `RecordingStart` and `RecordingStop`

**Comment by ii14** [comment]
You can already do that:

```vim
fun! s:record()
	" TODO: validate getchar()
	let s:last_reg = nr2char(getchar())
	execute 'normal! q'.s:last_reg
endfun

fun! s:exec_last()
	if s:last_reg !=# ''
		execute 'normal! @'.s:last_reg
	endif
endfun

noremap <expr> q reg_recording() ==# '' ? '<cmd>call <SID>record()<cr>' : 'q'
nnoremap Q <cmd>call <SID>exec_last()<cr>
```

**Comment by vigoux** [comment]
It feels hackish and not that well integrated ? After this it will be far easier :
```vim
nnoremap Q <cmd>execute 'normal @' .. reg_recorded()<CR>
```

**Comment by gpanders** [comment]
Rather than a default mapping, can we just change the default behavior of Q? Considering that 1) there is already `gQ` to do the same thing and 2) `Q` is a rarely used feature (in my anecdotal and unscientific opinion), why not get rid of the (admittedly small) overhead of a mapping and just go right to the source?

**Comment by vigoux** [comment]
I think we will be able to do all those using the PR mentionned, we can indeed change what `Q` does to repeat the last recorded macro, that's easy to do.

**Comment by gpanders** [comment]
> I think we will be able to do all those using the PR mentionned, we can indeed change what `Q` does to repeat the last recorded macro, that's easy to do.

My question was not "is it technically possible?" but "does it make sense to do this?". I know maintaining some continuity with Vim is important, but I'm wondering if it's important enough in this case to prevent us from changing the default action.

**Comment by vigoux** [comment]
I think it is fine to remove the old `Q` and use it for something else.
Everybody here agrees that having `Q` and `gq` do the same thing is useless.

**Comment by shadmansaleh** [comment]
Is it possible to make macros dot repeatable ?
Means `.` after recording a macro replays it

Or after replaying a macro (`@q` or `@@`) `.` replays it ?

`Q` mapping sounds great to . I doubt anyone is useing ex mode anymore .

**Comment by zeertzjq** [comment]
Implemented in #16684
